livepeer/go-livepeer
# Rudimentary failover

Still needs a lot more testing before I'm really comfortable with this, but wanted to get it out here for a quick look-over in case there are concerns about the general approach.

## What does this pull request do?
Implements rudimentary failover:

* Doesn't retry the current segment, but subsequent segments will select a new O.
* If a segment fails, lazily goes through the entire O selection process again rather than caching and using the list that was previously selected. This is actually pretty quick as long as at least one O responds, and is probably a better approach for long-running streams.
* We compare error strings to determine whether we should stop the session and failover. Not pretty, but works. Curious if there's a better approach. The list of error strings are compiled into a single regex for that extra oomph.

### Orchestrator Selector Listener

The major bit that is introduced here is the mechanism for setting the `rtmpConnection.sess` field within `startSessionListener`. A worker loop listens to the `rtmpConnection.needOrch` channel and kicks off the orchestrator-selection process. The orchestrator-selection process is mostly unchanged, but it's been extracted into its own function, `startSession`, with a few modifications for thread safety.

Some of this machinery may seem overwrought but it needs to be thread safe under a number of cases. Happy to walk through some of these if folks have questions, or if anyone has ideas for improvement.

The idea is to ensure exclusion of the `startSession` function: only one invocation can run at a time. If multiple requests for `startSession` come in while one is in-flight, they should be ignored. This prevents a pileup or unnecessary resetting of the session. This is accomplished with a traditional setup of a mutex and a condvar, but I'm also interested in exploring other approaches.

Implementation: https://github.com/livepeer/go-livepeer/commit/bc53ce29a2df4b7c3eb7eecc2fb166d43e7c2c56

An alternative is perhaps to do this with two channels, including a limited-length chan that rejects any attempts to write to it during an in-flight request. The transcode loop uses a similar method to indicate the busy state.

```
cxn.needOrch = make(chan struct{}, 1) // length limit of 1
cxn.getOrch = make(chan struct{}) // unlimited length?
cxn.eof = make(chan struct{})
...
if shouldStopSession(err) {
  select {
  case cxn.needOrch <- struct{}{}:
    cxn.getOrch <- struct{}{}
  case <-cxn.eof
     // rtmp stream ended; exit
     return
  case default:
    // already have an in flight request; exit
    return
  }
}
...
go func() {
  for {
    select {
    case <-cxn.getOrch:
      cxn.sess = nil
      cxn.sess = startSession() // select an O and get a session out of it
      <- cxn.needOrch // drain
    }
  }
}()
```

I'm still not convinced that we don't need locking somewhere. Putting on my old-school hat, I'm also not quite comfortable yet with the "golanginess"  of the approach. Explaining how this works would be extra thinking for folks that aren't familiar with golang concurrency constructs and the specific semantics thereof, such as the behavior of `select` when writing to a full chan. However, most developers should be comfortable with standard synchronization primitives.

## Specific updates

- Split out much of the logic in `gotRTMPStreamHandler` in standalone functions. This is finally straightforward thanks to the latest round of refactors that made the `rtmpConnection` struct available. Primarily `processSegment` and `startSession`.
- Place some code in `server/broadcast.go` to make `mediaserver.go` a bit less heavy. What got moved was pretty arbitrary, but in general, functions that aren't dependent on `LivepeerServer` go into `broadcast.go`

## Testing

1. Run a couple orchestrators in offchain mode. You'll want to set `-offchain` , `-serviceAddr`, `-cliAddr` and probably `-datadir` as well.
2. Run a broadcaster in `-offchain` mode with the above orchestrators specified in `-orchAddr orch1_ip:orch1_port,orch2_ip:orch2_port`
3. Start streaming into the broadcaster
4. Kill the active orchestrator that's working the stream
5. Ensure the second orchestrator picks it up
6. Restart the killed orchestrator
7. GOTO 4

Unit tests forthcoming.





**Does this pull request close any open issues?**



**Checklist:**


- [ ] Node runs in OSX and devenv
